[자동차 경주] Ornn 제출합니다.

SukWooJung opened the pull request:
재밌는 과제가 된 것 같네요.

저는 크게 **[자동차(Car)] - [자동차 관리(CarManager] - [게임(Game)]** 형태로 로직을 구분했습니다.

**로직**
입력을 통해 CarsName 받으면 이를 CarManager에 전달하여 생성과 함께 Car의 컬렉션 리스트 형태의 필드를 생성할 수 있도록, 생성자 의존 주입했습니다. 
CarManager에 존재하는 리스트형 필드인 cars를 통해 복수의 자동차 관리에 관한 로직을 모두 CarManager에서 담당할 수 있었습니다. 

자동차 관리와 게임의 역할은 다르다고 판단하여 Game 클래스를 생성했습니다.  Game에 참가할 CarManager를 마찬가지로 의존주입하여 Game 내에서 경기를 진행할 수 있도록 설정했습니다. 

**ps** 
- 입력에 관한 기능 및 익셉션 처리는 모두 InpuService에서 관리 했습니다. 
- 테스트는 유저가 잘못 입력하는 상황에 국한하여 만들었습니다. 
- 테스트 방식은 JUnitSpec을 통한 TDD와 BehaviorSpec을 통한 BDD를 만들었습니다.

**확인하고 학습할 사항**
1. 현재 실무의 코드에서도, 람다를 적극적으로 사용하는 코틀린의 특징을 적용시키는 시켜서 람다를 사용하는 지 아닌지, 
ex) for문 사용 or .forEach() 사용 

2. 사용하는 테스트 라이브러리는 어떤 것이며, 어떤 테스트 기법을 사용하는지 (BDD or TDD)

SukWooJung committed:
[Add] Car class 추가

SukWooJung committed:
feat(Car class): 필드 접근제어자 private 설정, 기본 생성자 X, position의 기본값은 0

SukWooJung committed:
build: KoTest 의존 추가

SukWooJung committed:
feat: 실행할 Main 파일 생성

SukWooJung committed:
feat: AnnotationSpec을 활용한 테스트파일 생성 및 차 이름 단위 테스트(TDD) 추가

SukWooJung committed:
feat: BehaviorSpec을 활용한 테스트파일 생성 및 차 이름 테스트(BDD) 추가

SukWooJung committed:
feat: 익셉션 추가

SukWooJung committed:
feat: 입력받는 역할을 하는 Service 추가

SukWooJung committed:
feat: 출력을 위해 만들어진, 문자열 소스 보관소 생성

SukWooJung committed:
style: 패키지명 소문자로 변경

SukWooJung committed:
feat: CarManager로 자동차 생성 및 몇 번 게임을 실행할지 입력받기

SukWooJung committed:
feat: 게임 시작 후 랜덤 하게 움직이고, 출력하는 기능 구현

SukWooJung committed:
feat: 게임 끝남을 알리는 문자열 추가

SukWooJung committed:
refact: 의존주입 방식을 setter에서 생성자 주입방식으로 변경
feat: 가장 멀리간 승리자를 찾아오는 함수 getWinners 구현

SukWooJung committed:
refact: Game 진행을 담당하는 객체 생성

SukWooJung committed:
refact: position 필드를 private set 형태로 변경 후 처리 (접근제어자 고민 이휴 해결)
feat: 이름을 출력하는 printName 함수 생성

SukWooJung committed:
refact: game 객체를 생성해서 코드 분리하는 작업을 진행

SukWooJung committed:
fix: 접근제어자 수정

SukWooJung committed:
feat: PlayingTimes 입력 시 발생하는 예외 테스트

SukWooJung committed:
refact: 불필요한 코드 중복 제거 (PlayingTimes)

perenok commented on build.gradle:
```diff
@@ -23,4 +25,4 @@ compileTestKotlin {
 test {
     useJUnitPlatform()
-}
+}
```
여기부터 여러 파일들에 아래 공백이 없는 것이 보이는데요ㅎㅎ
파일 마지막에 빈 줄을 넣는 이유에 대해 한번 찾아보시면 좋을 거 같아요!
그리고 빈 줄이 없는 채로 push하게 되면 깃허브에서 No newline at end of file이라고 경고가 뜹니다!
워터 pr에 피드백 달고보니 저도 막줄 경고가 수두룩... 호다닥 수정했습니다ㅋㅋㅋㅋ

perenok commented on src/main/kotlin/ornn/racingcar/data/Car.kt:
```diff
@@ -0,0 +1,33 @@
+            print("-")
+        }
+        println()
+    }
```
현재는 콘솔 출력 하나뿐이지만 추후 웹 View가 추가될 수 있고, 그럴 때 두 서비스에 보여지는 값이 다를 수 있어요. 
이렇게 View가 추가될 때 마다 도메인에 메서드가 늘어나는 것에 대해 고민해보아요!
위 고민을 통해 Domain과 View를 분리하는 것의 장점에 대해 같이 이야기 나눠봐요ㅋㅋㅋ

perenok commented on src/main/kotlin/ornn/racingcar/data/Car.kt:
```diff
@@ -0,0 +1,33 @@
+
+    fun printName(){
+        print(name)
+    }
```
view 로직이 도메인으로 넘어왔는데 name의 private을 없애서 getter를 활성화시키고 View에서 name을 가져가서 print로직을 구현하는 건 어떨까요?

perenok commented on src/main/kotlin/ornn/racingcar/data/Game.kt:
```diff
@@ -0,0 +1,35 @@
+            print(" ")
+            winner.printName()
+            if (winners.size == count) {
+                break;
```
```suggestion
                break
```
세미콜론에서 해방!

perenok commented on src/main/kotlin/ornn/racingcar/data/Game.kt:
```diff
@@ -0,0 +1,35 @@
+import ornn.racingcar.manager.CarManager
+import ornn.racingcar.resource.Strings
+
+class Game (private val carManager: CarManager, private val playingTimes : Int){
```
data 패키지를 저는 domain 혹은 model로 생각하고 코드를 읽었어요.
이를 기반으로 Car 클래스에서 언급한 내용처럼 도메인과 View를 분리해보면 좋을 거 같아요!

perenok commented on src/main/kotlin/ornn/racingcar/exception/IllegalInputException.kt:
```diff
@@ -0,0 +1,3 @@
+package ornn.racingcar.exception
+class IllegalInputException(override val message : String) : IllegalArgumentException(message)
```
커스텀 예외를 만드셨군요!
다만 현재는 예외의 이름이나 내부 구현에서 의도가 드러나는 게 없어서 IllegalArgumentException과 차이점을 잘 모르겠어요ㅠㅠ
어떤 예외를 잡으려는지 이름을 통해 의도를 더 드러내면 좋을 거 같아요ㅎㅎ

perenok commented on src/main/kotlin/ornn/racingcar/manager/CarManager.kt:
```diff
@@ -0,0 +1,41 @@
+
+import ornn.racingcar.data.Car
+
+class CarManager(carNames: List<String>) {
```
일급 컬렉션을 사용하셨군요! 👍

perenok commented on src/main/kotlin/ornn/racingcar/manager/CarManager.kt:
```diff
@@ -0,0 +1,41 @@
+        }
+    }
+
+    fun getWinners() : MutableList<Car> {
```
자체적으로 로직을 구현하셨는데 코틀린에서 제공하는 것들을 활용해도 좋을 거 같아요!

perenok commented on src/main/kotlin/ornn/racingcar/resource/Strings.kt:
```diff
@@ -0,0 +1,12 @@
+package ornn.racingcar.resource
+
+class Strings {
```
View에 필요한 상수들을 한 클래스에 모아뒀는데 이 경우와 각각을 Input, Output을 담당하는 클래스에서 관리하는 것과 어떤 차이가 있을까요?

perenok commented on src/main/kotlin/ornn/racingcar/service/InputService.kt:
```diff
@@ -0,0 +1,63 @@
+        while (true) {
+            println(Strings.ASK_CAR_NAMES)
+            val input = readLine().toString()
+            try {
+                checkInputIsNull(input)
+                carNames = input.split(",")
+                checkCarNameMoreThanFive(carNames)
+                break
```
try 문의 범위가 넓다면 어떠한 위치의 예외처리인지 파악이 힘들고 예상하지 못한 부분에서 catch 될 위험이 존재한다고 합니다!

perenok commented on src/test/kotlin/ornn/racingcar/ApplicationTestWithAnnotationSpec.kt:
```diff
@@ -0,0 +1,35 @@
+        val inputService = InputService()
+    }
+
+    // Kotest @AnnotationSpec
```
테스트 클래스 네이밍에도 나와 있고 상속을 통해서도 확인 가능해서 주석은 없어도 될 거 같은데 어떻게 생각하시나요?

perenok commented on src/test/kotlin/ornn/racingcar/ApplicationTestWithBehaviorSpec.kt:
```diff
@@ -0,0 +1,60 @@
+import ornn.racingcar.service.InputService
+import ornn.racingcar.exception.IllegalInputException
+
+class ApplicationTestWithBehaviorSpec : BehaviorSpec({
```
현재 테스트가 Application에서만 이루어지고 있어요!
단위테스트를 추가해서 작은 단위의 테스트가 있으면 좋을 거 같아요ㅎㅎ

perenok reviewed (commented):
안녕하세요 오른!
워터도 그렇고 다들 엄청 꼼꼼하게 예외처리를 하셨네요 대단합니다ㅋㅋㅋ
전 예외 신경도 안 쓰고 그냥 서비스 종료시켜버리네요ㅠㅠ 반성합니다
제 생각을 몇군데 남겨봤어요! 내일 같이 읽어보면서 의견 나누면 좋을 거 같습니다!
미션하느라 고생 많으셨어요ㅋㅋㅋ

osandra commented on src/main/kotlin/ornn/racingcar/data/Car.kt:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    fun move() {
+        val randomNum = Application.random.nextInt(MOVE_TO - MOVE_FROM) + MOVE_FROM
```
마지막에 MOVE_FROM 값인 0을 더해주는 부분은 생략하는 건 어떨까요?

osandra commented on src/main/kotlin/ornn/racingcar/service/InputService.kt:
```diff
@@ -0,0 +1,63 @@
+                break
+            } catch (e: IllegalInputException) {
+                println(Strings.ERROR_MESSAGE + " " + Strings.ERROR_INFO_INVALID_INPUT)
+            }
```
checkPlayTimeIsNum(input) 코드 아래에 try 문 안에서 진행되는 코드가 없어서 break 를 생략해도 좋을 것 같아요!

osandra commented on src/main/kotlin/ornn/racingcar/service/InputService.kt:
```diff
@@ -0,0 +1,63 @@
+            if (!c.isDigit()) {
+                throw IllegalInputException(Strings.ERROR_MESSAGE)
+            }
+        }
```
```suggestion
        try {
             input.toInt()
        } catch (e: NumberFormatException){
             throw IllegalInputException(Strings.ERROR_MESSAGE)
        }
```
이렇게 바꾸면 한 메서드 안에 indent(인덴트, 들여쓰기)가 1 이면서 원하는 결과를 얻을 수 있을 것 같아요 ㅎㅎ

osandra commented on src/main/kotlin/ornn/racingcar/Application.kt:
```diff
@@ -0,0 +1,25 @@
+import ornn.racingcar.data.Game
+import ornn.racingcar.manager.CarManager
+import ornn.racingcar.service.InputService
+import java.util.*
```
코틀린의 Random 패키지가 아니라 자바 패키지가 들어간 걸로 보이네요! 이 부분 코틀린 페키지로 변경해도 좋을 것 같아요 ㅎㅎ

osandra reviewed (approved):
안녕하세요 오른!
여러 테스트 방식을 사용하신 모습, 일급 컬렉션을 사용하시며 예외 처리한 모습 보면서 많이 배웠어요 ㅋㅋ 저도 일급 컬렉션을 공부하고 사용해봐야겠네요 🤩
보면서 든 생각들을 몇 개 적어봤어요! 
내일 같이 보면서 이야기해보아요 ~~ 고생 많으셨습니다!!

+ [확인하고 학습할 사항 1번]에 적어주신 내용처럼 실무에서도 람다를 사용하시더라고요! socar-server 레포 보면 사용하는 모습 보실 수 있어요! 저도 코틀린 눈에 익힐 겸 보고 있습니다 👀

socar-anakin commented on src/main/kotlin/ornn/racingcar/data/Car.kt:
```diff
@@ -0,0 +1,33 @@
+    fun printRecord() {
+        print("${name} : ")
+        for (i in 1..position) {
+            print("-")
```
`print()` 문을 사용하는 것보다는 logging을 사용하는게 좋습니다. 처음부터 이렇게 작성하는 습관을 들이시는게 좋습니다.

SukWooJung replied:
아나킨 코드리뷰 감사합니다. 혹시 Log 라이브러리는 어떤것을 사용하면 좋을까요 ??

socar-anakin commented on src/main/kotlin/ornn/racingcar/data/Car.kt:
```diff
@@ -0,0 +1,33 @@
+        }
+    }
+
+    fun printName(){
```
`(){` 로 붙혔는데, `() {`와 같이 한 칸 비워두세요. 작은 거고, Plugin을 사용하면 자동으로 처리되기도 하겠지만 좋은 습관이 몸에 배도록 하는게 중요합니다.

SukWooJung replied:
네 코드의 디테일한 부분도 신경써서 작성해야겠네요 감사합니다.

socar-anakin commented on src/main/kotlin/ornn/racingcar/data/Game.kt:
```diff
@@ -0,0 +1,35 @@
+            count++
+            print(" ")
+            winner.printName()
+            if (winners.size == count) {
```
불필요할 것 같은데?

socar-anakin replied:
연장선에서 `count`라는 것도 여기 `if` 문이 의미없어지면 함께 없애주는게 좋겠습니다.

socar-anakin commented on src/main/kotlin/ornn/racingcar/manager/CarManager.kt:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    fun printCarsRecord() {
+        for (car in cars) {
```
Stream을 사용하는게 더 좋을 것 같은데?

socar-anakin commented on src/main/kotlin/ornn/racingcar/service/InputService.kt:
```diff
@@ -0,0 +1,63 @@
+
+    fun readCarNames(): List<String> {
+        var carNames: List<String>
+        while (true) {
```
왜 while loop 이죠? 그리고 무한 루프는 안좋은 방식입니다.

socar-anakin commented on src/main/kotlin/ornn/racingcar/service/InputService.kt:
```diff
@@ -0,0 +1,63 @@
+        }
+    }
+
+    fun checkInputIsNull(input: String?) {
```
`check` 보다는 이 경우에 `validate` 라는 method naming이 더 맞습니다.

SukWooJung committed:
refact: 단위 테스트 추가, 테스트 파일 명 정정

SukWooJung committed:
refact: 발생한 에러에 대한 구체 메세지 추가

SukWooJung committed:
refact: 프로덕트 변경에 따른 메소드명 수정

SukWooJung committed:
feat: Random값을 도출하는 RandomGenerator를 띠로 정적 클래스 뺴서 생성

SukWooJung committed:
refactor: 패키지명을 domain으로 변경함.
Game 도메인 삭제 후 서비스 단인 CarService를 생성함
사유: Game 도메인 내에 저장될 데이터가 없음. 따로 관리가 안이루어진다.

SukWooJung committed:
refactor: CarManager 삭제 후 일급 컬렉션인 Cars 생성

SukWooJung committed:
refactor: 패키지명 domain으로 변경. printRecord 내부 스트림 함수로 처리

SukWooJung committed:
refactor: CarManager -> Cars 클래스명 변경. CarService를 생성하여 기능 분리

SukWooJung committed:
refactor: CarService를 생성 기능분리

SukWooJung committed:
refactor: while 문을 통한 무한루프 제거. 재귀 방식으로 변경. check => validate 메소드 명 변경

SukWooJung committed:
refactor: 프로덕션 변경에 따른 테스트 변경

SukWooJung committed:
refactor: 프로덕트 코드에 변경에 따라 수정. InputService, Random 정적으로 가져와서 사용

SukWooJung committed:
feat: Log 유틸 생성
